File: docs/breez-sdk/snippets/swift/BreezSdkSnippets/Sources/CrossChain.swift
import BigNumber
import BreezSdkSpark
import Foundation

func getCrossChainRoutes(sdk: BreezSdk) async throws {
    // ANCHOR: cross-chain-get-routes
    let input = "<recipient address>"
    let parsed = try await sdk.parse(input: input)
    guard case let .crossChainAddress(v1: addressDetails) = parsed else {
        throw NSError(domain: "CrossChain", code: 1)
    }

    let routes = try await sdk.getCrossChainRoutes(
        filter: .send(addressDetails: addressDetails))

    for route in routes {
        print("Route via \(route.provider): \(route.chain)/\(route.asset)")
        // Amount bounds are published per accepted asset. Read the ones for
        // the asset you intend to pay with, before quoting.
        for accepted in route.acceptedAssets {
            guard let limits = accepted.limits else {
                continue
            }
            print(
                "  \(accepted.asset) minimum: \(String(describing: limits.minAmount)) "
                    + "base units / \(String(describing: limits.minUsdCents)) USD cents"
            )
            if limits.dynamicLimitsPossible {
                print("  The provider may enforce a higher minimum than published")
            }
        }
    }
    // ANCHOR_END: cross-chain-get-routes
}

func prepareSendPaymentCrossChain(
    sdk: BreezSdk,
    addressDetails: CrossChainAddressDetails,
    route: CrossChainRoutePair
) async throws {
    // ANCHOR: cross-chain-prepare
    // Optionally set the maximum slippage in basis points (10 to 500)
    let optionalMaxSlippageBps: UInt32? = 100

    let prepareResponse = try await sdk.prepareSendPayment(
        request: PrepareSendPaymentRequest(
            paymentRequest: .crossChain(
                address: addressDetails.address,
                route: route,
                maxSlippageBps: optionalMaxSlippageBps,
                targetOverpayBps: nil
            ),
            amount: BInt(50_000),
            tokenIdentifier: nil,
            conversionOptions: nil,
            feePolicy: nil
        ))

    if case let .crossChainAddress(
        _, _, amountIn, _, estimatedOut, feeAmount, _, _, _, _, expiresAt, _
    ) = prepareResponse.paymentMethod {
        print("Amount in: \(amountIn)")
        print("Estimated out: \(estimatedOut)")
        print("Provider fee: \(feeAmount)")
        print("Quote expires at: \(expiresAt)")
    }
    // ANCHOR_END: cross-chain-prepare
}

func sendPaymentCrossChain(sdk: BreezSdk, prepareResponse: PrepareSendPaymentResponse) async throws
{
    // ANCHOR: cross-chain-send
    // Only valid for sends with no token leg (see Retry safety).
    let optionalIdempotencyKey = "<idempotency key uuid>"
    let sendResponse = try await sdk.sendPayment(
        request: SendPaymentRequest(
            prepareResponse: prepareResponse,
            options: nil,
            idempotencyKey: optionalIdempotencyKey
        ))
    let payment = sendResponse.payment
    print(payment)
    // ANCHOR_END: cross-chain-send
}

func getCrossChainReceiveRoutes(sdk: BreezSdk) async throws {
    // ANCHOR: cross-chain-get-receive-routes
    let routes = try await sdk.getCrossChainRoutes(
        filter: .receive(contractAddress: nil))

    for route in routes {
        print("Route via \(route.provider): \(route.chain)/\(route.asset) -> Spark")
    }
    // ANCHOR_END: cross-chain-get-receive-routes
}

func receivePaymentCrossChain(sdk: BreezSdk, route: CrossChainRoutePair) async throws {
    // ANCHOR: cross-chain-receive
    // amount is in the route's source-asset base units (USD-stable parity:
    // 1_000_000 = $1 on 6-decimal routes). See the guide for feeMode,
    // destination, and the slippage/overpay overrides.
    let amount = BInt(1_000_000)
    let optionalDestination: SparkAsset? = nil
    let optionalMaxSlippageBps: UInt32? = 100
    let optionalTargetOverpayBps: UInt32? = nil
    let optionalFeeMode: CrossChainFeeMode? = nil

    let response = try await sdk.receivePayment(
        request: ReceivePaymentRequest(
            paymentMethod: .crossChain(
                route: route,
                amount: amount,
                destination: optionalDestination,
                feeMode: optionalFeeMode,
                maxSlippageBps: optionalMaxSlippageBps,
                targetOverpayBps: optionalTargetOverpayBps
            )
        ))

    print("Payment request: \(response.paymentRequest)")
    if let info = response.crossChainInfo {
        print("Deposit address: \(info.depositAddress)")
        print("Deposit amount: \(info.depositAmount)")
        print(
            "Expected received: \(info.expectedReceivedAmount) "
                + "\(info.destinationAsset)"
        )
        print("Expires at: \(info.expiresAt)")
    }
    // ANCHOR_END: cross-chain-receive
}

File: docs/breez-sdk/snippets/flutter/lib/cross_chain.dart
import 'package:breez_sdk_spark_flutter/breez_sdk_spark.dart';

Future<List<CrossChainRoutePair>> getCrossChainRoutes(BreezSdk sdk) async {
  // ANCHOR: cross-chain-get-routes
  String input = "<recipient address>";
  InputType parsed = await sdk.parse(input: input);
  if (parsed is! InputType_CrossChainAddress) {
    throw Exception("Not a cross-chain address");
  }
  CrossChainAddressDetails addressDetails = parsed.field0;

  List<CrossChainRoutePair> routes = await sdk.getCrossChainRoutes(
    filter: CrossChainRouteFilter.send(addressDetails: addressDetails),
  );

  for (var route in routes) {
    print("Route via ${route.provider}: ${route.chain}/${route.asset}");
    // Amount bounds are published per accepted asset. Read the ones for
    // the asset you intend to pay with, before quoting.
    for (var accepted in route.acceptedAssets) {
      final limits = accepted.limits;
      if (limits == null) {
        continue;
      }
      print(
        "  ${accepted.asset} minimum: ${limits.minAmount} base units"
        " / ${limits.minUsdCents} USD cents",
      );
      if (limits.dynamicLimitsPossible) {
        print("  The provider may enforce a higher minimum than published");
      }
    }
  }
  // ANCHOR_END: cross-chain-get-routes
  return routes;
}

Future<PrepareSendPaymentResponse> prepareSendPaymentCrossChain(
  BreezSdk sdk,
  CrossChainAddressDetails addressDetails,
  CrossChainRoutePair route,
) async {
  // ANCHOR: cross-chain-prepare
  // Optionally set the maximum slippage in basis points (10 to 500)
  int? optionalMaxSlippageBps = 100;

  final request = PrepareSendPaymentRequest(
    paymentRequest: PaymentRequest.crossChain(
      address: addressDetails.address,
      route: route,
      maxSlippageBps: optionalMaxSlippageBps,
      targetOverpayBps: null,
    ),
    amount: BigInt.from(50000),
    tokenIdentifier: null,
    conversionOptions: null,
    feePolicy: null,
  );
  final response = await sdk.prepareSendPayment(request: request);

  final paymentMethod = response.paymentMethod;
  if (paymentMethod is SendPaymentMethod_CrossChainAddress) {
    print("Amount in: ${paymentMethod.amountIn}");
    print("Estimated out: ${paymentMethod.estimatedOut}");
    print("Provider fee: ${paymentMethod.feeAmount}");
    print("Quote expires at: ${paymentMethod.expiresAt}");
  }
  // ANCHOR_END: cross-chain-prepare
  return response;
}

Future<SendPaymentResponse> sendPaymentCrossChain(
  BreezSdk sdk,
  PrepareSendPaymentResponse prepareResponse,
) async {
  // ANCHOR: cross-chain-send
  // Only valid for sends with no token leg (see Retry safety).
  String? optionalIdempotencyKey = "<idempotency key uuid>";
  final request = SendPaymentRequest(
    prepareResponse: prepareResponse,
    options: null,
    idempotencyKey: optionalIdempotencyKey,
  );
  final response = await sdk.sendPayment(request: request);
  print("Payment: ${response.payment}");
  // ANCHOR_END: cross-chain-send
  return response;
}

Future<List<CrossChainRoutePair>> getCrossChainReceiveRoutes(BreezSdk sdk) async {
  // ANCHOR: cross-chain-get-receive-routes
  List<CrossChainRoutePair> routes = await sdk.getCrossChainRoutes(
    filter: CrossChainRouteFilter.receive(contractAddress: null),
  );

  for (var route in routes) {
    print(
      "Route via ${route.provider}: ${route.chain}/${route.asset} -> Spark",
    );
  }
  // ANCHOR_END: cross-chain-get-receive-routes
  return routes;
}

Future<ReceivePaymentResponse> receivePaymentCrossChain(
  BreezSdk sdk,
  CrossChainRoutePair route,
) async {
  // ANCHOR: cross-chain-receive
  // amount is in the route's source-asset base units (USD-stable parity:
  // 1_000_000 = $1 on 6-decimal routes). See the guide for feeMode,
  // destination, and the slippage/overpay overrides.
  final amount = BigInt.from(1000000);
  SparkAsset? optionalDestination;
  int? optionalMaxSlippageBps = 100;
  int? optionalTargetOverpayBps;
  CrossChainFeeMode? optionalFeeMode;

  final request = ReceivePaymentRequest(
    paymentMethod: ReceivePaymentMethod.crossChain(
      route: route,
      amount: amount,
      destination: optionalDestination,
      feeMode: optionalFeeMode,
      maxSlippageBps: optionalMaxSlippageBps,
      targetOverpayBps: optionalTargetOverpayBps,
    ),
  );
  final response = await sdk.receivePayment(request: request);

  print("Payment request: ${response.paymentRequest}");
  final info = response.crossChainInfo;
  if (info != null) {
    print("Deposit address: ${info.depositAddress}");
    print("Deposit amount: ${info.depositAmount}");
    print(
      "Expected received: ${info.expectedReceivedAmount} ${info.destinationAsset}",
    );
    print("Expires at: ${info.expiresAt}");
  }
  // ANCHOR_END: cross-chain-receive
  return response;
}
